adapters/repos/db/vector/cache/paginated_cache.go

```diff
@@ -0,0 +1,505 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package cache
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/common"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer"
+)
+
+// A PaginatedCache is a cache that stores vectors in pages of a fixed size.
+// Pages are created on demand and stored in a slice.
+// Each index in the slice can be either nil or point to a page.
+// The combination of slice and pages represent a virtual one-dimensional array of vectors.
+// Each vector is identified by its index in the virtual array.
+//
+// # Access and locking
+//
+// Reading and writing a single vector to the cache is a O(1) operation.
+// The vector id is used to calculate the page and index.
+// The cache uses a sharded lock to allow concurrent reads and writes,
+// however it is not possible to use the sharded lock to lock a single page.
+// This means that creating a page or increasing the size of the slice
+// requires a complete lock on the cache. This is not a problem in practice,
+// because these are relatively rare operations.
+//
+// # Memory usage and pruning
+//
+// When the cache slice is grown, pages are not allocated immediately.
+// Instead, pages are allocated when they are needed.
+// A pruning operation is run periodically to remove empty pages.
+type paginatedCache[T float32 | byte | uint64] struct {
+	shardedLocks     *common.ShardedLocks
+	cache            [][][]T
+	vectorForID      common.VectorForID[T]
+	maxSize          int64
+	count            int64
+	cancel           chan bool
+	logger           logrus.FieldLogger
+	deletionInterval time.Duration
+	pruneInterval    time.Duration
+
+	// The maintenanceLock makes sure that only one maintenance operation, such
+	// as growing the cache or clearing the cache happens at the same time.
+	maintenanceLock sync.RWMutex
+
+	// The pageLock makes sure that only one goroutine can create a page at the
+	// same time.
+	pageLock sync.Mutex
+}
+
+const (
+	// Size of a page in the cache
+	PageSize = 1024
+	// Number of pages to grow the cache by
+	GrowthDelta          = 6 * PageSize
```

Comment:
can be simply given as `6` to avoid converting it back again to number of pages in `Grow` method.